### ISSUE TITLE
End bone not show in the 3d view

### ISSUE THREAD

**Issue by firecloud888** [godotengine/godot#32393]
![image](https://user-images.githubusercontent.com/9168814/65780591-a14ec880-e17c-11e9-82c3-880942c0268b.png)

When import a mesh with a skeleton, you can't see the last bone in 3d view, the head bone or the finger and the leg bone.
By the way, seems the child mesh not follow the parent mesh if the parent is binding to a armature.

**Comment by marstaik** [comment]
What import file type are you using?

Are you exporting form blender? Blender does not export the tail of the last bone in the hierarchy (as its actually not a joint)

Can you provide sample files to test this and reproduce?

**Comment by fire** [comment]
Can anyone still reproduce this bug in Godot 3.2.3 or any later release?

If yes, please ensure that an up-to-date Minimal Reproduction Project (MRP) is included in this report (a MRP is a zipped Godot project with the minimal elements necessary to reliably trigger the bug). You can upload ZIP files in an issue comment with a drag and drop.

**Comment by MrMinimal** [comment]
@fire I'm showing the same behaviour in Godot 3.3.4.stable.official!

Relevant project (from #53466)
https://drive.google.com/file/d/1hL__pWpMKldgO745Uzs5g78lT5TlNctn/view?usp=sharing

![grafik](https://user-images.githubusercontent.com/8739690/136186449-4d73be5e-1fde-4b2a-acb3-f221dfcf60ee.png)

Vs Blender:
![grafik](https://user-images.githubusercontent.com/8739690/136186579-29b82188-d053-496f-ad4c-00a7033d7c0c.png)

**Comment by MrMinimal** [comment]
@TokageItLab You are right, using GLTF the imported Skeleton has the head bone (spine.006). But it gets ignored when generating the physical skeleton.

GLTF Skeleton with generated physical skeleton (spine.006 missing):
![grafik](https://user-images.githubusercontent.com/8739690/136334694-19d42e88-7272-422b-883e-1ff2a0dbba12.png)

Using the FBX importer, the head looks correct:
![grafik](https://user-images.githubusercontent.com/8739690/136333864-7e9e99b2-3193-4e1a-9c0b-7ac0d140bdd6.png)

Using the FBX Skeleton, the head bone does get a physical bone too:
![grafik](https://user-images.githubusercontent.com/8739690/136334754-b77e60fc-0886-484b-8f82-6186822c97df.png)

Shouldn't both formats produce the same result?

**Comment by Altair1234** [comment]
@TokageItLab Importing the gltf you exported from Blender right back into Blender has the end bones though. I'm not sure how this is supposed to work. It should probably also be mentioned somewhere (maybe in the documentation or somewhere people will see it) because I have seen quite a few people trip over this problem.

**Comment by rcorre** [comment]
> The workaround is to add leaf bones in Blender manually

@MrMinimal what do you mean by "manually"? Do you mean adding an "extra" leaf bone so the bone you actually want isn't a leaf bone any more?

**Comment by TokageItLab** [comment]
@rcorre I don't know what you are calling a "leaf bone".

To be exact, Godot doesn't have bones, Godot only has joints, and bones are just a visual concept to connect the joints. On the other hand, Blender has an object called a bone that has a head and a tail. Instead, it has no joints.

The glTF, like Godot, has only joints, and when Blender exports to glTF, it converts the position of the Head to the position of the Joint, and the vector and roll values from the Head to the Tail to the direction of the Joint. At this point, the Tail is discarded because the direction is not determined by the Tail alone.

I remember that fbx export has an option to add a "leaf bone" to the Tail to make it a bone with an appropriate direction, but glTF does not. So this is completely a blender problem, Godot can't solve it.

**Comment by TokageItLab** [comment]
@rcorre Your understanding of leaf bone is incorrect. Leaf bone is an unweighted bone with a fake orientation that exists only to help predict the length of the bone at the end of the chain in a jointed skeleton.

**Comment by rcorre** [comment]
Sorry if this is a silly idea ... but could we just guess? For a final bone in the chain (I don't know what to call this besides a "leaf"), could we just create a collider that is, say, 0.8 times the scale of its parent?

We might be wrong, but the scales for all the bones are just educated guesses anyways. Since they usually require manual tweaking after creation, I'd at least rather have the bones there. Having potentially wrong colliders feels less surprising than the bones being totally missing.

As a potentially crazier idea ... could we infer the bone size based on weights? If we formed an AABB enclosing the vertices influenced by a bone (possibly discarding those below some threshold), could that serve as a decent estimate of size?

**Comment by TokageItLab** [comment]
I think it is possible to add bones by prediction as an option of CreatePhysicsBone. However, I'm strongly opposed to having an option in EditorSetting to display the predicted bones.

**Comment by TokageItLab** [comment]
@rcorre For reference, VRM 0 as the glTF extension format's SpringBone, used virtual colliders in extend 0.07m of end joint, but it will be removed in VRM 1.0. This is just my guess, but I think the reason is that the prediction specification was hard to understand for users.

**Comment by rcorre** [comment]
Thanks, that's a good point of reference.

My assumption is that "the PhysicalBone is totally missing for a bone I know exists" is more suprising to the end user than "The PhysicalBone exists but the collider needs to be tweaked", especially given the latter is often the case even for non-leaf bones.

Even https://docs.godotengine.org/en/stable/tutorials/physics/ragdoll_system.html#collision-shape-adjustment suggests that you'll need to adjust collision shapes after creation. It doesn't suggest that you'll need to add bones that are missing.

**Comment by TokageItLab** [comment]
I think the adjustment of the collision shape is due to the fact that the bones have no thickness. As long as the collider is not generated on the end joint, it does not mention with regards to length.
It should definitely be documented regarding the missing tail of the end bone in Blender with a sentence about "the blender lacking a function".

**Comment by MrMinimal** [comment]
If Godot is set up for `.blend` file import, it imports them without the "Add Leaf Bones" option enabled. I think it should be default behaviour to import the armature as you see it in Blender.

<img width="250" height="455" alt="Image" src="https://github.com/user-attachments/assets/767bf9d0-3ba7-4273-9173-5598b3bb3e00" />

@TokageItLab what do you think?
